man/man1/textile.1.md
% TEXTILE(1)
% Jukka Svahn
% July 2025

# NAME

textile -- Compile Textile markup documents to HTML

# SYNOPSIS

**textile** [*options*] `<`*file*`>`

# DESCRIPTION

CLI implementation of PHP-Textile, a modern Textile markup language parser for PHP. Textile is a humane web
text generator that takes lightweight, readable, plaintext-like markup language and converts it into well-formed
HTML.

# OPTIONS

`-h`, `--help`
: Print help.

`-V`, `--version`
: Print version.

`-d` `<`*type*`>`, `--document-type`[=]=`<`*type*`>`
: Specifies output document type that is generated from the given Textile input. Supported values are `html5` and
`xhtml`. Defaults to `html5`.

`--document-root-directory`[=]=`<`*directory*`>`
: Sets the directory path that is used to resolve relative file paths within local filesystem. For instance, this is
used for resolving image dimensions. If not set, defaults to the current working directory.

`-l`, `--lite`
: Enables lite mode. If enabled, allowed tags are limited. Parser will prevent the use of extra
Textile formatting, accepting only paragraphs and blockquotes as valid block tags.

: The option doesn't prevent the use of potentially unsafe document features. If you wish to parse untrusted user-given
Textile input, also enable the restricted mode with `--restricted` or `-r` option.

`-i`, `--no-images`
: Disables images. If disabled, image tags are not generated. This option can be used for minimalist output such as
text-only documents.

`-L`, `--link-relationship`
: Sets the HTML `rel` attribute value that are applied to links generated from Textile input. Good values could
include **nofollow** or **noreferrer**.

`-r`, `--restricted`
: Enables restricted mode. This option should be enabled when parsing untrusted user input, including comments or
social media posts. When enabled, the parser escapes any raw HTML input, ignores unsafe attributes and generates links
only for whitelisted safe URL schemes.

: If restricted mode is left disabled, the parser allows users to mix raw HTML and Textile markup. Using the parser in
non-restricted on untrusted input, like comments and forum posts, will lead to XSS issues, as users will be able to use
any HTML code, JavaScript links and Textile attributes in their markup input.

`-U`, `--raw-blocks`
: Enables raw blocks. When raw blocks are enabled, any paragraph blocks wrapped in a non-standard HTML or XML like tags
will be left as is, without parsing the paragraph's content. Recognized phrasing tags and block HTML tags are
excluded.

`-A`, `--align-classes`
: Force enables alignment classes. By default, in when using **html5** as the target `--document-type`, HTML img
elements are generated with **align-left**, **align-center** and **align-right** class selectors rather than align
attribute being added to the element. With this option you can enable that functionality for the **xhtml** document type
mode too.

`-a`, `--no-align-classes`
: Force disables alignment classes. Generated **html5** document will also end up using invalid align attributes,
instead of classes.

`-b`, `--no-block-tags`
: Disables rendering of block tags. The given Textile input will not be wrapped to paragraph blocks, allowing the
formatting of a single line of text.

`-w`, `--no-line-wrap`
: Disables line wrapping. Input document's line-breaks are ignored. This setting can be used if the input document's
lines are pre-wrapped. For instance, in case the input is from CLI content, or source code documentation.

`-p`, `--image-prefix`
: Sets image URL prefix. The given URL is used to prefix relative image paths.

`-P`, `--link-prefix`
: Sets link URL prefix. The given URL is used to prefix relative link paths.

`-z`, `--no-dimensions`
: Disable adding **width** and **height** dimensions to images.

`-o` `<`*file*`>`, `--output`[=]=`<`*file*`>`
: Path to where the resulting output HTML document is saved to. If not given, the results are printed to
**STDOUT**.

`-q`, `--quiet`
: Do not output any messages.

`--ansi`
: Force enable ANSI output.

`--no-ansi`
: Force disable ANSI output.

`--no-interaction`
: Do not ask any interactive question.

`-v`, `-vv`, `-vvv`, `--verbose`
: Increase verbosity. `-v` for verbose, `-vv` for more verbose and `-vvv` for debug.

*file*
: Input Textile document file to process. Alternatively, the input can be read from `STDIN`.

# ENVIRONMENT

`NO_COLOR`
: Variable is used to figure whether output should use colors.

# FILES

Depends on external programs `php` as the runtime. Requires PHP version 8.1 or newer.

# EXAMPLES

Source Textile document can be specified with an argument, and the location
where the HTML output documented is saved to with `--output` option:

    $ textile --output=generated.html source.textile

Compiles piped standard input containing Textile markup into HTML:

    $ echo "Hello world" | textile

The `--restricted` flag can be used when processing untrusted user-input, it
will disable features that could generate potentially dangerous HTML:

    $ textile --restricted path/to/untrusted.textile


CHANGELOG.md
# Changelog

## 0.1.1

* Re-release of 0.1.0. Fixed issue in release CI.

## 0.1.0

* Initial release.


src/Command/Textile.php
<?php

declare(strict_types=1);

/*
 * PHP-Textile CLI
 * https://github.com/gocom/php-textile-cli
 *
 * Copyright (C) 2025 Jukka Svahn
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

namespace Rah\TextileCli\Command;

use Netcarver\Textile\Parser;
use Rah\TextileCli\Input\GetStdInAction;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Input\StreamableInputInterface;
use Symfony\Component\Console\Output\ConsoleOutputInterface;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * Parse command.
 *
 * Compiles Textile documents into HTML.
 */
final class Textile extends Command
{
    /**
     * {@inheritdoc}
     */
    protected function configure(): void
    {
        $this
            ->setName('textile')
            ->setDefinition([
                new InputOption(
                    'document-type',
                    't',
                    InputOption::VALUE_REQUIRED,
                    'Set the document type',
                    Parser::DOCTYPE_HTML5,
                    [
                        Parser::DOCTYPE_HTML5,
                        Parser::DOCTYPE_XHTML,
                    ]
                ),

                new InputOption(
                    'document-root-directory',
                    'd',
                    InputOption::VALUE_REQUIRED,
                    'Set the document root directory',
                    ''
                ),

                new InputOption(
                    'lite',
                    'l',
                    InputOption::VALUE_NONE,
                    'Enable lite mode'
                ),

                new InputOption(
                    'no-images',
                    'i',
                    InputOption::VALUE_NONE,
                    'Disable images'
                ),

                new InputOption(
                    'link-relationship',
                    'L',
                    InputOption::VALUE_REQUIRED,
                    'Set link relationship',
                    null,
                    ['nofollow', 'noreferrer']
                ),

                new InputOption(
                    'restricted',
                    'r',
                    InputOption::VALUE_NONE,
                    'Enable restricted mode'
                ),

                new InputOption(
                    'raw-blocks',
                    'U',
                    InputOption::VALUE_NONE,
                    'Enable raw HTML blocks'
                ),

                new InputOption(
                    'align-classes',
                    'A',
                    InputOption::VALUE_NONE,
                    'Enable alignment classes'
                ),

                new InputOption(
                    'no-align-classes',
                    'a',
                    InputOption::VALUE_NONE,
                    'Disable alignment classes'
                ),

                new InputOption(
                    'no-block-tags',
                    'b',
                    InputOption::VALUE_NONE,
                    'Disable block tags'
                ),

                new InputOption(
                    'no-line-wrap',
                    'w',
                    InputOption::VALUE_NONE,
                    'Disable line wrapping'
                ),

                new InputOption(
                    'image-prefix',
                    'p',
                    InputOption::VALUE_REQUIRED,
                    'Set image URL prefix'
                ),

                new InputOption(
                    'link-prefix',
                    'P',
                    InputOption::VALUE_REQUIRED,
                    'Set link URL prefix'
                ),

                new InputOption(
                    'no-dimensions',
                    'z',
                    InputOption::VALUE_NONE,
                    'Disable adding width and height to images'
                ),

                new InputOption(
                    'output',
                    'o',
                    InputOption::VALUE_REQUIRED,
                    'Save output HTML as the specified file'
                ),

                new InputArgument(
                    'file',
                    InputArgument::OPTIONAL,
                    'Textile document to parse and compile'
                ),
            ])
            ->setDescription('Compile Textile markup to HTML')
            ->setHelp(<<<EOF
CLI implementation of PHP-Textile, a modern Textile markup language parser for PHP. Textile is a humane web
text generator that takes lightweight, readable, plaintext-like markup language and converts it into well-formed
HTML.
EOF
            );
    }

    /**
     * {@inheritdoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $errorOutput = $output instanceof ConsoleOutputInterface
            ? $output->getErrorOutput()
            : $output;

        $inputFile = $input->getArgument('file');

        if ($inputFile) {
            if (!\file_exists($inputFile) || !\is_readable($inputFile)) {
                $errorOutput->writeln(
                    '<error>Textile document file could not be opened.</error>'
                );

                return Command::FAILURE;
            }

            $document = \file_get_contents($inputFile);
        } else {
            $inputSteam = $input instanceof StreamableInputInterface
                ? $input->getStream()
                : null;

            $document = (new GetStdInAction())->execute($inputSteam);
        }

        if (!$document) {
            $errorOutput->writeln(
                '<error>Textile document is required.</error>'
            );

            return Command::FAILURE;
        }

        $textile = new Parser();

        $textile
            ->setDocumentType($input->getOption('document-type') ?? Parser::DOCTYPE_HTML5)
            ->setLite((bool) $input->getOption('lite'))
            ->setImages(!$input->getOption('no-images'))
            ->setLinkRelationShip((string) $input->getOption('link-relationship'))
            ->setRestricted((bool) $input->getOption('restricted'))
            ->setRawBlocks((bool) $input->getOption('raw-blocks'))
            ->setBlockTags(!$input->getOption('no-block-tags'))
            ->setLineWrap(!$input->getOption('no-line-wrap'))
            ->setImagePrefix((string) $input->getOption('image-prefix'))
            ->setLinkPrefix((string) $input->getOption('link-prefix'))
            ->setDimensionlessImages((bool) $input->getOption('no-dimensions'));

        if ($input->getOption('document-root-directory')) {
            $textile->setDocumentRootDirectory($input->getOption('document-root-directory'));
        }

        if ($input->getOption('align-classes')) {
            $textile->setAlignClasses(true);
        }

        if ($input->getOption('no-align-classes')) {
            $textile->setAlignClasses(false);
        }

        $html = $textile->parse($document);
        $saveAs = $input->getOption('output');

        if ($saveAs) {
            $result = \file_put_contents(
                $saveAs,
                $html . "\n"
            );

            if ($result === false) {
                $errorOutput->writeln(
                    \sprintf(
                        '<error>Failed to save HTML document as %s</error>',
                        $saveAs
                    )
                );

                return Command::FAILURE;
            }

            return Command::SUCCESS;
        }

        $output->writeln($html);

        return Command::SUCCESS;
    }
}
